### PR TITLE
feat(ms-cryptoapi): add support for MS CryptoAPI tokens

### PR DESCRIPTION
**Pull request by mrts**
WE2-716

### CODE REVIEW

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    SECURITY_STATUS err = ERROR_SUCCESS;
+    std::wstring unicodePin = toUnicode(pin);
+    err = NCryptSetProperty(key, NCRYPT_PIN_PROPERTY, PBYTE(unicodePin.c_str()),
+                            DWORD(unicodePin.size()), 0);
+    switch (err) {
+    case SCARD_W_WRONG_CHV:
+    case SCARD_E_INVALID_CHV:
+    case ERROR_INVALID_PARAMETER:
+        // TODO: find a way to avoid relying on PIN_RETRY_COUNT_PLACEHOLDER.
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.cpp:76]: It is probably not possible to find the PIN retry count using current CNG APIs.

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            THROW(WrongCertificateTypeError,
+                  "This electronic ID does not support " + std::string(typ) + " certificates");
+        }
+        return certData;
+    }
+
+    JsonWebSignatureAlgorithm authSignatureAlgorithm() const override
+    {
+        // TODO: PS256
+        return isRSA() ? JsonWebSignatureAlgorithm::RS256 : JsonWebSignatureAlgorithm::ES384;
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp:81]: ES384 is not good, find out the curve length from the certificate public key.

File: src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+using namespace electronic_id;
+
+struct CertStatus
+{
+    bool isValid = false;
+    bool isAuthentication = false;
+};
+
+// Checks if the provided certificate time is valid and key usage is for signing or authentication.
+CertStatus getCertStatus(PCCERT_CONTEXT cert)
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp:45]: Use OpenSSL version from PKCS#11 instead?

File: src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    };
+
+    PCCERT_CONTEXT cert = nullptr;
+    while ((cert = CertEnumCertificatesInStore(sys, cert)) != nullptr) {
+        auto certStatus = getCertStatus(cert);
+        if (!certStatus.isValid) {
+            continue;
+        }
+        DWORD flags = CRYPT_ACQUIRE_CACHE_FLAG | CRYPT_ACQUIRE_COMPARE_KEY_FLAG
+            | CRYPT_ACQUIRE_SILENT_FLAG | CRYPT_ACQUIRE_PREFER_NCRYPT_KEY_FLAG;
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp:124]:
```suggestion
            | CRYPT_ACQUIRE_SILENT_FLAG | CRYPT_ACQUIRE_REQUIRE_NCRYPT_KEY_FLAG;
```

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    virtual ~MsCryptoApiElectronicID()
+    {
+        if (freeKey) {
+            NCryptFreeObject(key);
+        }
+        CertFreeCertificateContext(certContext);
+    }
+
+    static const int8_t PIN_RETRY_COUNT_PLACEHOLDER = -1;
+    static const int8_t PIN_MIN_LENGTH_PLACEHOLDER = 3;
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp:59]:
```suggestion
    static const int8_t PIN_MIN_LENGTH_PLACEHOLDER = 4;
```

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    static const int8_t PIN_MAX_LENGTH_PLACEHOLDER = 100;
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp:60]:
```suggestion
    static const int8_t PIN_MAX_LENGTH_PLACEHOLDER = 16;
```

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+              "EVP_PKEY_base_id() reports non-EC key where EC key expected");
+    }
+
+    auto keyBitLength = EVP_PKEY_bits(key);
+    switch (keyBitLength) {
+    case 256:
+        return JsonWebSignatureAlgorithm::ES256;
+    case 384:
+        return JsonWebSignatureAlgorithm::ES384;
+    case 512:
```
Comment (mrts) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.cpp:56]:
```suggestion
    case 512:
    case 521:
```

File: src/availableSupportedCards.cpp [hunk 8]
```diff
@@ -45,11 +49,9 @@ std::vector<CardInfo::ptr> availableSupportedCards()
         readers = pcsc_cpp::listReaders();
         std::vector<CardInfo::ptr> cards;
 
-        if (readers.empty()) {
-            throw AutoSelectFailed(AutoSelectFailed::Reason::NO_READERS);
-        }
-
         auto seenCard = false;
+        // The list may be empty, but we cannot throw yet due to the availableMSCryptoApiTokens()
```
Comment (mrts) [inline src/availableSupportedCards.cpp:53]:
```suggestion
        // The list may be empty, but we cannot throw yet due to the listMsCryptoApiElectronicIDs()
```

File: src/electronic-ids/x509.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+{
+
+inline static void* extension(X509* x509, int nid)
+{
+    return X509_get_ext_d2i(x509, nid, nullptr, nullptr);
+}
+
+inline static bool hasClientAuthExtendedKeyUsage(EXTENDED_KEY_USAGE* usage)
+{
+
```
Comment (mrts) [inline src/electronic-ids/x509.hpp:20]:
```suggestion
```

File: src/electronic-ids/x509.hpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+#include "scope.hpp"
+
+#include <openssl/x509v3.h>
+#include <openssl/err.h>
+
+namespace electronic_id
+{
+
+inline static void* extension(X509* x509, int nid)
```
Comment (mrts) [inline src/electronic-ids/x509.hpp:13]:
```suggestion
inline void* extension(X509* x509, int nid)
```

File: src/electronic-ids/x509.hpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+{
+    return X509_get_ext_d2i(x509, nid, nullptr, nullptr);
+}
+
+inline static bool hasClientAuthExtendedKeyUsage(EXTENDED_KEY_USAGE* usage)
```
Comment (mrts) [inline src/electronic-ids/x509.hpp:18]:
```suggestion
inline bool hasClientAuthExtendedKeyUsage(EXTENDED_KEY_USAGE* usage)
```

File: src/availableSupportedCards.cpp [hunk 12]
```diff
@@ -60,8 +62,22 @@ std::vector<CardInfo::ptr> availableSupportedCards()
...
+#ifdef _WIN32
+        // In Windows, also include CryptoAPI tokens.
+        // Initially, CryptoAPI tokens will be included only if there are
+        // no tier 1/2 cards present.
+        if (cards.empty()) {
+            auto msCryptoApiElectronicIDs = listMsCryptoApiElectronicIDs();
+            if (!msCryptoApiElectronicIDs.empty()) {
+                seenCard = true;
+                cards = msCryptoApiElectronicIDs;
+            }
```
Comment (metsma) [inline src/availableSupportedCards.cpp:74]:
```suggestion
            cards = listMsCryptoApiElectronicIDs();
```
seenCard is already previously set true

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    std::string name() const override
+    {
+        // TODO: use NCryptGetProperty(key, NCRYPT_NAME_PROPERTY, ...).
+        return "MS CryptoAPI electronic ID";
+    }
+    Type type() const override { return Type::MsCryptoApiEID; }
+
+    bool isRSA() const { return signatureAlgo != SignatureAlgorithm::ES; }
+
+    const PCCERT_CONTEXT certContext;
```
Comment (metsma) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp:119]:
```suggestion
   PCCERT_CONTEXT certContext;
```
Type name indicates already const

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+                            bool freeK) :
+        ElectronicID {std::make_unique<pcsc_cpp::SmartCard>()},
+        certContext {certCtx}, certData {cert}, certType {cType},
+        // TODO: SignatureAlgorithm::PS?
+        signatureAlgo {isRsa ? SignatureAlgorithm::RS : SignatureAlgorithm::ES}, key {k}, freeKey {
+                                                                                              freeK}
+    {
+    }
+
+    virtual ~MsCryptoApiElectronicID()
```
Comment (metsma) [inline src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp:50]:
```suggestion
   ~MsCryptoApiElectronicID()
```
Missing copy&move constructors and operators

File: src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        // The certificate context must also remain alive to use the key, so call
+        // CertDuplicateCertificateContext() to increment its reference count to keep it alive.
+        // MsCryptoApiElectronicID frees the certificate context itself.
+        auto eid = std::make_unique<MsCryptoApiElectronicID>(CertDuplicateCertificateContext(cert),
+                                                             std::move(certData), certType,
+                                                             algo == L"RSA", key, freeKey);
+
+        msCryptoApiElectronicIDs.emplace_back(
```
Comment (metsma) [inline src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp:150]:
```suggestion
        msCryptoApiElectronicIDs.push_back(
```

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp (hunk 14 above)
Comment (mrts) [reply src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp]: Thanks, good point!

File: src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp (hunk 13 above)
Comment (mrts) [reply src/electronic-ids/ms-cryptoapi/MsCryptoApiElectronicID.hpp]: Thanks, fixed.

File: src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp (hunk 15 above)
Comment (mrts) [reply src/electronic-ids/ms-cryptoapi/listMsCryptoApiElectronicIDs.cpp:150]: `emplace_back` is preferrable according to https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-emplace.html, but there are indeed other opinions as well like https://quuxplusone.github.io/blog/2021/03/03/push-back-emplace-back/. Why do you prefer `push_back`?

**Pull request merged**